// amis/esm/renderers/Status.js
/**
 * amis v2.7.2
 * Copyright 2018-2023 baidu
 */

import { __extends, __decorate } from 'tslib';
import React from 'react';
import merge from 'lodash/merge';
import { getPropValue, filter, Renderer } from 'amis-core';
import { Icon } from 'amis-ui';

var StatusField = /** @class */ (function (_super) {
    __extends(StatusField, _super);
    function StatusField() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StatusField.prototype.render = function () {
        var _a, _b;
        var _c = this.props, className = _c.className, style = _c.style, placeholder = _c.placeholder, cx = _c.classnames, data = _c.data;
        var map = merge(StatusField.defaultProps.map, (_a = this.props) === null || _a === void 0 ? void 0 : _a.map);
        var labelMap = merge(StatusField.defaultProps.labelMap, (_b = this.props) === null || _b === void 0 ? void 0 : _b.labelMap);
        var value = getPropValue(this.props);
        var viewValue = (React.createElement("span", { className: "text-muted", key: "status-value" }, placeholder));
        var wrapClassName = '';
        if (value != undefined && value !== '' && map) {
            if (typeof value === 'boolean') {
                value = value ? 1 : 0;
            }
            else if (/^\d+$/.test(value)) {
                value = parseInt(value, 10) || 0;
            }
            wrapClassName = "StatusField--".concat(value);
            var itemClassName = filter(map[value], data) || '';
            var svgIcon_1 = '';
            itemClassName = itemClassName.replace(/\bsvg-([^\s|$]+)\b/g, function (_, icon) {
                svgIcon_1 = icon;
                return 'icon';
            });
            if (svgIcon_1) {
                viewValue = (React.createElement(Icon, { icon: svgIcon_1, className: cx('Status-icon icon', itemClassName), key: "icon" }));
            }
            else if (itemClassName) {
                viewValue = (React.createElement("i", { className: cx('Status-icon', itemClassName), key: "icon" }));
            }
            if (labelMap && labelMap[value]) {
                viewValue = [
                    viewValue,
                    React.createElement("span", { className: cx('StatusField-label'), key: "label" }, filter(labelMap[value], data))
                ];
            }
        }
        return (React.createElement("span", { className: cx('StatusField', wrapClassName, className), style: style }, viewValue));
    };
    StatusField.defaultProps = {
        placeholder: '-',
        map: {
            0: 'svg-fail',
            1: 'svg-success',
            success: 'svg-success',
            pending: 'rolling',
            fail: 'svg-fail',
            queue: 'svg-warning',
            schedule: 'svg-schedule'
        },
        labelMap: {
            success: '成功',
            pending: '运行中',
            fail: '失败',
            queue: '排队中',
            schedule: '调度中'
        }
    };
    return StatusField;
}(React.Component));
/** @class */ ((function (_super) {
    __extends(StatusFieldRenderer, _super);
    function StatusFieldRenderer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StatusFieldRenderer = __decorate([
        Renderer({
            type: 'status'
        })
    ], StatusFieldRenderer);
    return StatusFieldRenderer;
})(StatusField));

export { StatusField };
