// amis-ui/lib/hooks/use-validation-resolver.js
/**
 * amis-ui v2.7.2
 * Copyright 2018-2023 fex
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var pick = require('lodash/pick');
var React = require('react');
var amisCore = require('amis-core');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var pick__default = /*#__PURE__*/_interopDefaultLegacy(pick);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function formatErrors(errors) {
    var formated = {};
    Object.keys(errors).forEach(function (key) {
        var origin = errors[key][0];
        if (origin) {
            formated[key] = {
                type: origin.rule,
                message: origin.msg
            };
        }
    });
    return formated;
}
function useValidationResolver(__, validate) {
    var _this = this;
    if (__ === void 0) { __ = function (str) { return str; }; }
    return React__default["default"].useCallback(function (values, context, config) { return tslib.__awaiter(_this, void 0, void 0, function () {
        var rules, customValidator, ruleKeys, _i, _a, key, field, errors, _b, _c, key, validate_1, result, e_1;
        return tslib.__generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    rules = {};
                    customValidator = {};
                    ruleKeys = Object.keys(amisCore.validations);
                    for (_i = 0, _a = Object.keys(config.fields); _i < _a.length; _i++) {
                        key = _a[_i];
                        field = config.fields[key];
                        rules[key] = pick__default["default"](field, ruleKeys);
                        if (field.required) {
                            rules[key].isRequired = true;
                        }
                        if (typeof field.validate === 'function') {
                            customValidator[key] = field.validate;
                        }
                    }
                    errors = amisCore.validateObject(values, rules, undefined, __);
                    _b = 0, _c = Object.keys(customValidator);
                    _d.label = 1;
                case 1:
                    if (!(_b < _c.length)) return [3 /*break*/, 4];
                    key = _c[_b];
                    validate_1 = customValidator[key];
                    return [4 /*yield*/, validate_1(values[key])];
                case 2:
                    result = _d.sent();
                    if (typeof result === 'string') {
                        errors[key] = errors[key] || [];
                        errors[key].push({
                            rule: 'custom',
                            msg: result
                        });
                    }
                    _d.label = 3;
                case 3:
                    _b++;
                    return [3 /*break*/, 1];
                case 4:
                    _d.trys.push([4, 6, , 7]);
                    return [4 /*yield*/, (validate === null || validate === void 0 ? void 0 : validate(errors, values, context, config))];
                case 5:
                    _d.sent();
                    return [3 /*break*/, 7];
                case 6:
                    e_1 = _d.sent();
                    errors.customValidate = [
                        {
                            rule: 'custom',
                            msg: e_1.message || e_1
                        }
                    ];
                    return [3 /*break*/, 7];
                case 7: return [2 /*return*/, {
                        values: values,
                        errors: formatErrors(errors)
                    }];
            }
        });
    }); }, [__, validate]);
}

exports["default"] = useValidationResolver;
exports.useValidationResolver = useValidationResolver;
